keps/sig-scheduling/5055-dra-device-taints-and-tolerations/README.md

```diff
@@ -399,34 +429,109 @@ As with Taint, these structs get duplicated to document DRA specific
 behavior and to ensure that future extensions do not get inherited
 accidentally.
 
-Generated conversion code might make it possible to reuse existing helper
-code. Alternatively, that code can be copied.
-
-The DevicePatch also gets extended. It is possible to use
-admin-controlled taints without enabling attribute overrides by enabling the
-`v1alpha3` API and only the `DRADeviceTaints` feature, while leaving
-`DRAAdminControlledDeviceAttributes` disabled, because then the
-ResourceSlicePatch type is available with only the fields needed for
-taints.
+The implementation of the taint logic gets copied from node taints:
+[resourceclaim.ToleratesTaint](https://github.com/kubernetes/kubernetes/blob/85734ac6b38e29a9a390520e7a5b6de1fbf5ff6b/staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/devicetoleration.go#L23-L53)
+corresponds to
+[v1.Toleration.ToleratesTaint](https://github.com/kubernetes/kubernetes/blob/85734ac6b38e29a9a390520e7a5b6de1fbf5ff6b/staging/src/k8s.io/api/core/v1/toleration.go#L29-L57).
+
+To enable setting of taints without reconfiguring DRA drivers, the
+cluster-scoped DeviceTaintRule gets added. The scheduler must merge these
+additional taints with the ones provided by
+the DRA drivers on-the-fly while it gathers information about available
+devices. This could be done each time a scheduling cycle starts, but it
+would be repetitive work. Instead, a
+[ResourceSlice tracker](https://github.com/kubernetes/kubernetes/blob/85734ac6b38e29a9a390520e7a5b6de1fbf5ff6b/staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker.go#L56-L59)
+reacts to informer events for ResourceSlice and DeviceTaintRule and
+maintains a set of updated ResourceSlices which also contain the taints
+set via a DeviceTaintRule. The tracker provides the API of an informer
+and thus can be used as a drop-in replacement for a ResourceSlice
+informer.
```

Comment:
Why do we handle `DeviceTaintRule` in the scheduler? Is there any reason we cannot have a DeviceTaintRule controller keep watching DeviceTaintRules and apply taints to Device accordingly? In that way, other components (including scheduler) don't have to be aware of `DeviceTaintRule`, reducing complexity.

Reply:
That alternative had been considered: https://github.com/kubernetes/enhancements/tree/d7a10fa4c07c75538ae24977e5a680ae8db63840/keps/sig-scheduling/5027-dra-admin-controlled-device-attributes#storing-result-of-patching-in-resourceslice

Let me copy that here.

Reply:
=> https://github.com/kubernetes/enhancements/compare/d7a10fa4c07c75538ae24977e5a680ae8db63840..951be37a26964117143c9232fa496f5818983090

Reply:
And https://github.com/kubernetes/enhancements/compare/951be37a26964117143c9232fa496f5818983090..4a18dfbbf414162f0691ac711fc0f76ba78ae3d8.

Sorry, should have double-checked what I copied...